From d4b99a76d45d69b79c91c192f0b94608fa565d07 Mon Sep 17 00:00:00 2001
From: Jes Sorensen <Jes.Sorensen@redhat.com>
Date: Mon, 1 Sep 2014 09:25:21 -0400
Subject: [md] raid5: add proper locking to error path of raid5_start_reshape

Message-id: <1409563584-9000-20-git-send-email-Jes.Sorensen@redhat.com>
Patchwork-id: 89981
O-Subject: [RHEL7.1 PATCH 19/82] md/raid5.c: add proper locking to error path of raid5_start_reshape.
Bugzilla: 1085530
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

From: NeilBrown <neilb@suse.de>

If raid5_start_reshape errors out, we need to reset all the fields
that were updated (not just some), and need to use the seq_counter
to ensure make_request() doesn't use an inconsitent state.

Signed-off-by: NeilBrown <neilb@suse.de>
(cherry picked from commit ba8805b97320416e7c5bb8f55d2bd06d5c319e7d)
Signed-off-by: Jes Sorensen <Jes.Sorensen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/md/raid5.c b/drivers/md/raid5.c
index 0c7c311..a65ebee 100644
--- a/drivers/md/raid5.c
+++ b/drivers/md/raid5.c
@@ -6372,12 +6372,18 @@ static int raid5_start_reshape(struct mddev *mddev)
  if (!mddev->sync_thread) {
   mddev->recovery = 0;
   spin_lock_irq(&conf->device_lock);
+  write_seqcount_begin(&conf->gen_lock);
   mddev->raid_disks = conf->raid_disks = conf->previous_raid_disks;
+  mddev->new_chunk_sectors =
+   conf->chunk_sectors = conf->prev_chunk_sectors;
+  mddev->new_layout = conf->algorithm = conf->prev_algo;
   rdev_for_each(rdev, mddev)
    rdev->new_data_offset = rdev->data_offset;
   smp_wmb();
+  conf->generation --;
   conf->reshape_progress = MaxSector;
   mddev->reshape_position = MaxSector;
+  write_seqcount_end(&conf->gen_lock);
   spin_unlock_irq(&conf->device_lock);
   return -EAGAIN;
  }
-- 
1.7.1